### PR TITLE
relation: proof-linking: Add group placement over roots of unity

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR refactors the implementation of proof-linking arithmetization to have an `alignment` which suggests the roots-of-unity that the proof linking gates should be placed on. This allows proofs of different domain sizes to link with one another.

### Testing
- All unit tests pass

### CODE REVIEW

File: relation/src/proof_linking/arithmetization.rs [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+        // 1. Place the proof linking groups with specific layouts into the circuit
+        let alignment = self.circuit_alignment();
+        let mut sorted_placements = self
+            .link_group_layouts
+            .clone()
+            .into_iter()
+            .sorted_by_key(|(_, layout)| layout.range_in_nth_roots(alignment))
+            .collect_vec();
+
+        // 2. Place the rest of the gates into the circuit
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs]:
```suggestion
        // 2. Place the rest of the link groups into the circuit
```

File: relation/src/proof_linking/arithmetization.rs [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+    /// That is, for return value `n`, the circuit will be aligned to the 2^n-th
+    /// roots of unity
+    fn circuit_alignment(&self) -> usize {
+        let max_alignment =
+            self.link_group_layouts.values().map(|l| l.alignment).max().unwrap_or(0);
+
+        let n_proof_links = self.link_groups.values().map(|v| v.len()).sum::<usize>();
+        let n_gates = self.num_gates() + n_proof_links;
+
+        usize::max(n_gates.ilog2() as usize, max_alignment)
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs]: I'm pretty sure the domain is generated by taking the next power of two from the number of gates, but `ilog` will round down to the previous power of two.

Also, will this circuit alignment necessarily give enough space? if the total number of spacing "slots" introduced by the link groups is greater than `self.num_gates()`, then there will be more than `n_gates` gates in the final circuit. This can be the case even when `max_alignment < n.gates.ilog2()`, since `max_alignment` here only comes from the the alignments of the groups which have a predetermined layout (those in `link_group_layouts`), and not from all of the final alignments given by `gen_circuit_layout`.

Unless, however, this function is meant to give the smallest possible, but not necessarily final, circuit alignment (which seems to be how its used in `place_group`), in which case I strongly suggest renaming it and making sure we use `CircuitLayout::circuit_size().ilog2()` where necessary as opposed to this

File: relation/src/proof_linking/arithmetization.rs [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+    /// Place a group into a list of already allocated groups
+    ///
+    /// Assumes the `placed_groups` are sorted by their range
+    ///
+    /// Mutates `placed_groups` to insert the new group
+    ///
+    /// We first attempt to place the group in the largest set of roots of unity
+    /// that already has a group allocated. If successful, this optimally
+    /// uses the spread of the roots of unity (larger sets are more compactly
+    /// spaced), and does not require an increase in proof size. If
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs]: IIUC, this is saying that if we can place the new group within the largest alignment that's already necessitated by an existing group (which is the smallest alignment that the whole circuit could use), then we can avoid increasing the total circuit alignment and thus spreading out the previously allocated group.

If my understanding is correct, could we add that to the doc comment? It's a little hard to understand at first how "largest" and "smallest" are used in the doc comment vs the code comments in the method

File: relation/src/proof_linking/arithmetization.rs [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+                return true;
+            } else {
+                // We've found a gap that's too small, so we skip over it
+                offset = end + 1;
+                curr_idx += 1;
+            }
+        }
+
+        // If we've reached the end of the placed groups, we attempt to place the group
+        // at the end of the roots of unity
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs]: wouldn't it be more accurate to say that we're attempting to place the group directly after the already-placed groups? this would read more clearly

File: relation/src/constraint_system.rs [hunk 5]
```diff
@@ -80,8 +81,36 @@ pub struct CircuitLayout {
...
+impl CircuitLayout {
+    /// Get the domain size used to represent the circuit after proof linking
+    /// gates are accounted for
+    pub fn circuit_size(&self) -> usize {
+        // Check for link group alignments that need larger roots
+        let max_alignment =
+            self.group_layouts.values().map(|layout| layout.alignment).max().unwrap_or(1);
+        let link_gates = self.group_layouts.iter().map(|(_, layout)| layout.size).sum::<usize>();
+
+        let gates = (self.n_gates + link_gates).next_power_of_two();
```
Comment (akirillo) [inline relation/src/constraint_system.rs:97]: shouldn't we first check if the sum is a power of 2 before bumping it to the next one (this is how the domain size is determined, I believe)?

File: relation/src/constraint_system.rs [hunk 6]
```diff
@@ -80,8 +81,36 @@ pub struct CircuitLayout {
...
+        let alignment = 1 << max_alignment;
```
Comment (akirillo) [inline relation/src/constraint_system.rs]: nit: can we call this `max_required_roots_of_unity`, or something other than `alignment`, as it confuses the term?

File: relation/src/constraint_system.rs [hunk 7]
```diff
@@ -80,8 +81,36 @@ pub struct CircuitLayout {
...
+        usize::max(gates, alignment)
+    }
+
+    /// Get an iterator over the groups in the circuit, sorted by the ranges
+    /// they are allocated in
+    ///
+    /// Group placement guarantees these ranges are disjoint, so it is safe to
+    /// sort them in any common alignment
```
Comment (akirillo) [inline relation/src/constraint_system.rs]: nit: this isn't necessarily true, since we use this method in validating the group placement. Could we maybe reword this to "We assume the groups were placed in a disjoint manner"?

File: relation/src/proof_linking/arithmetization.rs [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+        let gates_iter = &mut old_gates;
+        let vars_iter = &mut old_vars;
+
+        // 3. Place the input gates from the gates iter
+        Self::place_gates(layout.n_inputs, gates, wire_variables, gates_iter, vars_iter);
+
+        // 4. Place the proof linking gates
+        let circuit_alignment = self.circuit_alignment();
+        for (id, layout) in layout.sorted_groups_iter() {
+            // Insert gates until the start of the next group
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs]:
```suggestion
            // Insert padding gates until the start of the next group
```

File: relation/src/proof_linking/arithmetization.rs [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+        let new_vars = vars_iter.iter_mut().map(|v| v.chain(iter::repeat(0)).take(n)).collect_vec();
+
+        gates.extend(&mut new_gates);
+        for (wire, var) in wire_variables.iter_mut().zip(new_vars) {
+            wire.extend(var);
+        }
+    }
+
+    /// Place a group in the circuit interspersed with gates from the original
+    /// circuit
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs:283]: very nice, I wasn't sure if we would intersperse w/ circuit gates at first but props for implementing this

File: relation/src/proof_linking/arithmetization.rs [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+        let gates = &mut new_gates;
+        let wire_variables = &mut new_wire_variables;
+        let gates_iter = &mut old_gates;
+        let vars_iter = &mut old_vars;
+
+        // 3. Place the input gates from the gates iter
+        Self::place_gates(layout.n_inputs, gates, wire_variables, gates_iter, vars_iter);
+
+        // 4. Place the proof linking gates
+        let circuit_alignment = self.circuit_alignment();
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs]: should this be `layout.circuit_size().ilog2() as usize`?

File: relation/src/proof_linking/arithmetization.rs [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+        &self,
+        id: &str,
+        layout: &GroupLayout,
+        gates: &mut Vec<Box<dyn Gate<F>>>,
+        wire_variables: &mut [Vec<Variable>; GATE_WIDTH + 2],
+        gates_iter: &mut impl Iterator<Item = Box<dyn Gate<F>>>,
+        vars_iter: &mut Vec<impl Iterator<Item = Variable>>,
+    ) {
+        // Translate the layout alignment to the circuit's alignment
+        let spacing = 1 << (self.circuit_alignment() - layout.alignment);
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs]: should this as well be `layout.circuit_size().ilog2() - layout.alignment`?

File: relation/src/constraint_system.rs [hunk 12]
```diff
@@ -174,9 +203,9 @@ where
     // placement of groups in the arithmetization
     /// The proof-linking group layouts for the circuit. Maps a group ID to the
     /// indices of the witness contained in the group
-    link_groups: HashMap<String, Vec<Variable>>,
-    /// The offset of each group in the arithmetization
-    link_group_offsets: HashMap<String, usize>,
+    pub(crate) link_groups: HashMap<String, Vec<Variable>>,
+    /// The layouts of each link group in the arithmetization
```
Comment (akirillo) [inline relation/src/constraint_system.rs]: IIRC not every link group will have a `GroupLayout`, only those that come from other circuits, right? If so, can we mention that in this doc comment (as opposed to "each link group in the arithmetization")?

**Review by akirillo — COMMENTED** [review]
Left a bunch of readability nits since this is understandable a tough implementation to wrap one's head around. But I do want to be sure that `PlonkCircuit::circuit_alignment()` is correctly implemented/used before approving

File: relation/src/proof_linking/arithmetization.rs (hunk 2 above)
Comment (joeykraut) [reply relation/src/proof_linking/arithmetization.rs]: discussed offline, for posterity: we decided to use a `circuit_alignment` method on the `CircuitLayout` which guarantees that all group placements have been finalized. This leaves enough space.

File: relation/src/constraint_system.rs (hunk 5 above)
Comment (joeykraut) [reply relation/src/constraint_system.rs:97]: The implementation of `next_power_of_two` returns the input if the input is a power of two

Comment (akirillo) [reply relation/src/constraint_system.rs:97]: :0

File: relation/src/proof_linking/arithmetization.rs [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+        }
+
+        // Otherwise, we've exhausted all possible placements
+        false
+    }
+
+    /// Validate a circuit layout against the circuit's arithmetization
+    fn validate_layout(&self, layout: &CircuitLayout) -> Result<(), CircuitError> {
+        // Check that each offset occurs after the public inputs
+        let n = layout.circuit_size().ilog2() as usize;
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs]: nit: use the fancy new `layout.circuit_alignment()` method you just created

File: relation/src/proof_linking/arithmetization.rs [hunk 14]
```diff
@@ -0,0 +1,480 @@
...
+
+        // Group 3
+        let placement3 = layout.group_layouts.get("group3").unwrap();
+        assert_eq!(placement3.alignment, alignment3);
+        assert_eq!(placement3.offset, offset3);
+        assert_eq!(placement3.size, 1);
+
+        // Check that the link groups have been placed correctly
+        cs.finalize_for_arithmetization().unwrap();
+        assert_eq!(cs.current_circuit_alignment(), alignment3);
```
Comment (akirillo) [inline relation/src/proof_linking/arithmetization.rs]: should this now be `layout.circuit_alignment()`?

**Review by akirillo — APPROVED** [review]
just a couple last nits, nice work here
